src/metabase/api/card.clj

```diff
@@ -703,6 +706,126 @@
                  :js-int-to-string?      false}))
 
 
+;;; ------------------------------------------------ Parameters -------------------------------------------------
+(def ^:const result-limit
+  "How many results to return when chain filtering."
+  1000)
+
+(defn- field-clause->field-id
+  "Find the field id in a field clause if it exists.
+
+    (field-clause->field-id [:field 3 nil])
+    ;; -> 3"
+  [field-clause]
+  (mbql.u/match-one field-clause [:field (id :guard integer?) _] id))
+
+(defn- template-tag-target->field-id
+  "Given a template tag target, find the field id that it's connected to.
+  Note that a target has a field id connect to it iff it's a field filter."
+  [card {:keys [id]}]
+  (if-let [template-tags (get-in card [:dataset_query :native :template-tags])]
+    (first (for [[_ template-tag] template-tags
+                  :when (and (= (:id template-tag) id)
+                             (= (:type template-tag) :dimension))
+                  :let  [field-id (field-clause->field-id (:dimension template-tag))]
+                  :when field-id]
+             field-id))
+    (throw (ex-info (tru "Card with ID {0} does not have template tags." (:id card))
+                    {:card-id         (:id card)
+                     :template-tag-id id}))))
+
+(defn- target->field-id
+  "Find the field id that the target is connected to.
+  Target could be a `dimension`, in which target have the shape
+
+    [:dimension [:field 3 nil]]
+
+  Or it could be a `template-tag`, then it should have the shape
+
+    [:template-tag {:id \"6006ad7d-036e-83ec-4d6f-30f82b98ac21\"}]"
+  [card [target-type target-args]]
+  (case target-type
+    :dimension    (field-clause->field-id target-args)
+    :template-tag (template-tag-target->field-id card target-args)
+    nil))
+
+(defn- param-key->field-ids
+ "Get Field ID(s) associated with a parameter in a Card.
+
+    (param-key->field-ids (Card 62) \"ee876336\")
+    ;; -> #{276}"
+  [{:keys [parameter_mappings] :as card} param-key]
+  (into #{} (for [{:keys [target parameter_id]} parameter_mappings
+                  :when (= parameter_id param-key)
+                  :let [field-id (target->field-id card target)]
+                  :when field-id]
+              field-id)))
+
+(defn- chain-filter-constraints [card constraint-param-key->value]
+  (into {} (for [[param-key value] constraint-param-key->value
+                 field-id          (param-key->field-ids card param-key)]
+             [field-id value])))
+
+(s/defn chain-filter
+  "C H A I N filters!
+
+    ;; show me categories
+    (chain-filter (Card 62) \"ee876336\" {})
+    ;; -> (\"African\" \"American\" \"Artisan\" ...)
+
+    ;; show me categories that have expensive restaurants
+    (chain-filter (Card 62) \"ee876336\" {\"6f10a41f\" 4})
+    ;; -> (\"Japanese\" \"Steakhouse\")"
```

Comment:
Chain filters are (at least currently) a dashboard-specific concept, so I don't think it makes sense to mention it/support it for cards, right? 

We can still use the `chain-filter` ns for the underlying implementation, but just only support looking up values with one filter at a time via the API.

Reply:
It would also simplify this code a lot to remove the chain filter-specific stuff.

Reply:
Let's me review what I think chain-filter does when we give it a `field`:
1. figure out if this field is cached in FieldValues, return if it is
2. otherwise, fetch from DB
3. Apply human-readable mapping if it exists
4. on top of the 3 steps above, make sure it follows a set of constraints if that exists

I don't think this is a dashboard-specific concept other than the 4th point.

Also, most of the code here is to find the field-id from parameter and parameter_mappings, which I think regardless of whether or not we use `chain-filter`, we still need to do this.

That said, I don't see a shorter way to do this without losing all the magic that `chain-filter` is giving us.
Or maybe I'm missing something?